tools/shared.py

```diff
@@ -2974,16 +2984,61 @@ def lebify(x):
     return bytearray(ret)
 
   @staticmethod
-  def make_shared_library(js_file, wasm_file, needed_dynlibs):
-    # a wasm shared library has a special "dylink" section, see tools-conventions repo
+  def get_js_data(js_file, shared=False):
     js = open(js_file).read()
     m = re.search("var STATIC_BUMP = (\d+);", js)
     mem_size = int(m.group(1))
     m = re.search("Module\['wasmTableSize'\] = (\d+);", js)
     table_size = int(m.group(1))
-    m = re.search('gb = alignMemory\(getMemory\(\d+ \+ (\d+)\), (\d+) \|\| 1\);', js)
-    assert m.group(1) == m.group(2), 'js must contain a clear alignment for the wasm shared library'
-    mem_align = int(m.group(1))
+    if shared:
+      m = re.search('gb = alignMemory\(getMemory\(\d+ \+ (\d+)\), (\d+) \|\| 1\);', js)
+      assert m.group(1) == m.group(2), 'js must contain a clear alignment for the wasm shared library'
+      mem_align = int(m.group(1))
+    else:
+      mem_align = None
+    return (mem_size, table_size, mem_align)
+
+  @staticmethod
+  def add_emscripten_metadata(js_file, wasm_file):
+    (mem_size, table_size, _) = WebAssembly.get_js_data(js_file)
+    wasm = open(wasm_file, 'rb').read()
+
+    raw_name = b'emscripten_metadata'
+    name = bytes(bytearray([len(raw_name)])) + raw_name
+    contents = b''.join(map(bytes, map(WebAssembly.lebify, [
+      # metadata section version
+      EMSCRIPTEN_METADATA_MAJOR,
+      EMSCRIPTEN_METADATA_MINOR,
+      # NB: The structure of the following should only be changed
+      #     if EMSCRIPTEN_METADATA_MAJOR is incremented
+      # ABI version
+      EMSCRIPTEN_ABI_MAJOR,
+      EMSCRIPTEN_ABI_MINOR,
+      # static bump
+      mem_size,
+      # table size
+      table_size,
+      # NB: more data can be appended here as long as you increase
+      #     the EMSCRIPTEN_METADATA_MINOR
+    ])))
+
+    new_section = b''.join([
+      b'\0', # custom section byte ID (== 0)
+      bytes(WebAssembly.lebify(len(name) + len(contents))),
+      name,
+      contents,
+    ])
+
+    with open(wasm_file, 'wb') as f:
```

Comment:
the style of this code differs from the `dylink` emitting code nearby (in minor details like using `with`, and other structuring). if we think this is nicer, which seems reasonable, let's change the other one too.

Reply:
That's fair, I can make it mirror the `dylink` emitting code.